eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/DPState.hs

```diff
@@ -235,19 +268,166 @@ instance Default (DState c) where
       Map.empty
       (GenDelegs Map.empty)
       def
+      Map.empty
 
 instance Default (PState c) where
   def =
-    PState Map.empty Map.empty Map.empty
+    PState Map.empty Map.empty Map.empty Map.empty
 
 rewards :: DState c -> ViewMap c (Credential 'Staking c) Coin
-rewards (DState unified _ _ _) = Rewards unified
+rewards (DState unified _ _ _ _) = Rewards unified
 
 delegations ::
   DState c ->
   ViewMap c (Credential 'Staking c) (KeyHash 'StakePool c)
-delegations (DState unified _ _ _) = Delegations unified
+delegations (DState unified _ _ _ _) = Delegations unified
 
 -- | get the actual ptrs map, we don't need a view
 ptrsMap :: DState c -> Map Ptr (Credential 'Staking c)
-ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _) = ptrmap
+ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _ _) = ptrmap
+
+-- ==========================================================
+-- Functions that handle Deposits for stake credetials and key hashes.
+
+-- | One only pays a deposit on the initial key registration. If the key has been
+--   de-registered it should have been removed from the map. If it hasn't been
+--   de-registered, then it has no effect on the Deposits. In places where this function
+--   is called, there should be an explicit check that the credential is not in the map.
```

Comment:
I don't understand exactly this: `In places where this function is called, there should be an explicit check that the credential is not in the map.`  For example, `Deleg.delegationTransition` calls this function,  so it's expected that it does a check before calling it?  
I can see this check: 
` eval (hk ∉ dom (rewards ds)) ?! StakeKeyAlreadyRegisteredDELEG hk` , so should it always check in the rewards map ?

Reply:
It is not that simple, because when we process a block, the bock can have multiple DCerts, so it is possible to Register and the Trregister in the dame block. The test you point to is just the state before the block. It might have changed as we process the mutiple DCerts in the block.  Maybe I can make the comment clearer.